### PR TITLE
[FLINK-27199][Connector/Pulsar] Bump pulsar to 2.10.0

### PR DESCRIPTION
**Pull request by syhily**
## What is the purpose of the change

Bump the Pulsar to 2.10.0 release which fixes a lot of issues on Pulsar transaction, especially the long waiting in Unit tests.

## Brief change log

  - Bump pulsar dependencies to 2.10.0
  - Fix the Pulsar mock instance which drops the zookeeper.

## Verifying this change

This change is already covered by existing tests, such as All the Pulsar connector's unit tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): **(yes)**
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

**Comment by MartijnVisser** [comment]
@flinkbot run azure

**Comment by syhily** [comment]
@MartijnVisser This PR is ready and gets CI passed. Can you invite someone to review it?

**Comment by syhily** [comment]
@codelipenghui Can you help me review this PR to see if I'm correctly using the memory meta store and testing on the latest 2.10.0 Pulsar?

File: flink-connectors/flink-connector-pulsar/src/test/resources/docker/bootstrap.sh [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+# ----------------------------------------------------------------------------
+
+# Enable the transaction in standalone config.
+sed -i 's/transactionCoordinatorEnabled=false/transactionCoordinatorEnabled=true/g' /pulsar/conf/standalone.conf
+sed -i 's/acknowledgmentAtBatchIndexLevelEnabled=false/acknowledgmentAtBatchIndexLevelEnabled=true/g' /pulsar/conf/standalone.conf
+sed -i 's/systemTopicEnabled=false/systemTopicEnabled=true/g' /pulsar/conf/standalone.conf
```
Comment (codelipenghui) [inline flink-connectors/flink-connector-pulsar/src/test/resources/docker/bootstrap.sh:24]: For enabling the transaction, it needs to enable `brokerDeduplicationEnabled=true` as well. The transaction depends on deduplication to avoid duplicated messages while the producer resends the messages due to connection issues or broker crashes.

**Review by codelipenghui — APPROVED** [review]
LGTM, just left a minor comment.

File: flink-connectors/flink-connector-pulsar/src/test/resources/docker/bootstrap.sh (hunk 1 above)
Comment (syhily) [reply flink-connectors/flink-connector-pulsar/src/test/resources/docker/bootstrap.sh:24]: OK, I'll add it.

**Review by BewareMyPower — APPROVED** [review]

**Comment by MartijnVisser** [comment]
@syhily The CI hasn't passed for this PR, please have a look

File: docs/layouts/shortcodes/generated/pulsar_client_configuration.html [hunk 2]
```diff
@@ -100,7 +100,7 @@
         </tr>
         <tr>
             <td><h5>pulsar.client.memoryLimitBytes</h5></td>
-            <td style="word-wrap: break-word;">0</td>
```
Comment (fapaul) [inline docs/layouts/shortcodes/generated/pulsar_client_configuration.html:103]: I think this change deserves at least a separate commit. Is this changing the behavior or correcting the docs?

File: flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/sink/PulsarSinkOptions.java [hunk 3]
```diff
@@ -128,6 +127,16 @@ private PulsarSinkOptions() {
...
                                     + " This is used in Pulsar Transaction.");
 
+    public static final ConfigOption<Integer> PULSAR_MAX_PENDING_MESSAGES_ON_PARALLELISM =
+            ConfigOptions.key(SINK_CONFIG_PREFIX + "maxPendingMessages")
+                    .intType()
+                    .defaultValue(1000)
+                    .withDescription(
+                            Description.builder()
+                                    .text(
+                                            "The maximum number of pending messages in on sink parallelism.")
```
Comment (fapaul) [inline flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/sink/PulsarSinkOptions.java:137]: Can you explain this configuration value? The current description is hard to understand.

File: flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/sink/PulsarSinkOptions.java [hunk 4]
```diff
@@ -128,6 +127,16 @@ private PulsarSinkOptions() {
                             "The allowed transaction recommit times if we meet some retryable exception."
                                     + " This is used in Pulsar Transaction.");
 
+    public static final ConfigOption<Integer> PULSAR_MAX_PENDING_MESSAGES_ON_PARALLELISM =
```
Comment (fapaul) [inline flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/sink/PulsarSinkOptions.java:130]: Can you add this configuration on a separate commit?

File: flink-connectors/flink-connector-pulsar/src/test/java/org/apache/flink/connector/pulsar/testutils/runtime/PulsarRuntime.java [hunk 5]
```diff
@@ -50,22 +49,10 @@ static PulsarRuntime mock() {
...
 
-    /**
-     * Create a standalone Pulsar instance in test thread. We would start an embedded zookeeper and
-     * bookkeeper. The stream storage for bookkeeper is disabled. The function worker is disabled on
-     * Pulsar broker.
-     *
-     * <p>This runtime would be faster than {@link #container()} and behaves the same as the {@link
-     * #container()}.
-     */
-    static PulsarRuntime embedded() {
```
Comment (fapaul) [inline flink-connectors/flink-connector-pulsar/src/test/java/org/apache/flink/connector/pulsar/testutils/runtime/PulsarRuntime.java:61]: Has this something to do with the version bump? If not, please use a separate commit to remove the embedded test environment.

**Review by fapaul — COMMENTED** [review]
The version bump looks good, and I also verified that the necessary NOTICE files have been touched, but please split the PR in multiple commits. Currently, it seems the one commit is pretty overloaded and does not provide enough context.

File: docs/layouts/shortcodes/generated/pulsar_client_configuration.html (hunk 2 above)
Comment (syhily) [reply docs/layouts/shortcodes/generated/pulsar_client_configuration.html:103]: Yep

File: flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/sink/PulsarSinkOptions.java (hunk 3 above)
Comment (syhily) [reply flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/sink/PulsarSinkOptions.java:137]: OK

**Comment by fapaul** [comment]
@syhily, what is the current state of this PR? I know you have investigated some other failures in the mean, but the version bump also seems to be quite critical for the stability.

**Comment by syhily** [comment]
@fapaul This is PR ready, I think.

**Comment by MartijnVisser** [comment]
@syhily How quickly is it possible to improve this PR? If I'm not mistaken, this is the most important PR to get in since it should resolve the current test instabilities. In the current state that the Pulsar connector is in, it's threatening the stability of the Flink 1.16 release.

**Comment by syhily** [comment]
> Thanks for the patch @syhily. I notice that this PR touches a lot of components including Pulsar client version, docs even default value of sink options. What about breaking these changes into different commits or PRs? I'm afraid squashing them into one commit is way too ambiguous.

OK, I'll split this PRs.

**Comment by syhily** [comment]
@PatrickRen I have splitted this PR into multiple commits.

**Review by PatrickRen — COMMENTED** [review]
Thanks for the update @syhily . The patch overall looks good to me except 2 typos I guess.

**Comment by MartijnVisser** [comment]
> The patch overall looks good to me except 2 typos I guess.

Are you OK with merging this when these typos have been addressed? Would be great if we can get it merged today.

**Comment by syhily** [comment]
@PatrickRen The typo has been fixed.
